### ISSUE TITLE
Rule: require functions to have a return statement

### ISSUE THREAD

**Issue by modernserf** [eslint/eslint#913]
Functions implicitly return `undefined` if no return statement is provided. However, this can sometimes mask errors, for example:

``` js
function increment(x){
    x + 1;
}
```

This was intended to return x + 1, but instead returns undefined.

This rule is aimed at ensuring all normal functions have at least one explicit exit; i.e. a `return` or `throw` statement. Constructors and IIFEs do not need a return as they are not expected to return a value.

I'm using this rule because I've encountered a number of cases where I have simply forgot to return a value. This is particularly bothersome with async code that uses Promises, which reads like it works on side effects but actually requires the Promise to be returned.

These are dumb and obvious mistakes, of course, but catching dumb and obvious mistakes is exactly the job of a linter.

This rule is not for everybody (hence off by default) but can be useful for those in circumstances such as myself, or when one wants to study how much of their code is relying on side-effects.
